Created scripts for benchmarking with wiredtiger

JacksonDagger opened the pull request:
Tools in the bench directory to benchmark PIM-swap with wiredtiger. qemu-test.sh is the file used to run buildroot on a QEMU guest. parse-logs.py is used to parse the logs to turn them into meaningful data. Commit also updates swap_setup.sh and .gitignore

JacksonDagger committed:
Created scripts for benchmarking with wiredtiger

Tools in the bench directory to benchmark PIM-swap with wiredtiger. qemu-test.sh is the file used to run buildroot on a QEMU guest. parse-logs.py is used to parse the logs to turn them into meaningful data. Commit also updates swap_setup.sh and .gitignore

JacksonDagger commented:
I made a mess of git so I squashed all the commits. The commits (with incorrect dates due to rebase) are all on the benchmarking branch. Other than my own benchmarking stuff, the only other thing I edited was .gitignore swap_setup.sh so it can take the swap file size as a command line arg. However, if you run swap_setup.sh normally, it will still work and the default is still 1G

jnider commented on bench/qemu-test.sh:
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
I don't like that there are two scripts for running qemu (run-qemu.sh and /bench/qemu-test.sh) that have a lot of overlap. Is it possible to merge these into a single script?

JacksonDagger replied:
yes. I think I can rework it

JacksonDagger replied:
run-qemu.sh is now deleted. sudo ./qemu-test.sh (with no CL args) should do the same thing as run-qemu

jnider commented on bench/qemu-test.sh:
```diff
@@ -0,0 +1,58 @@
+TERMINAL="gnome-terminal"
+
+# command line arguments to script with defaults
+HOST_SH="${1:-wiredtiger-copy.sh}" # file to run on host
```
This looks like it is specific to wired tiger right now. What do we need to change to make this generic? We will have several benchmarks to run eventually.

JacksonDagger replied:
yeah I'll fix that along with making it optional

JacksonDagger replied:
fixed now. By default, there is no guest or host script

jnider commented on bench/qemu-test.sh:
```diff
@@ -0,0 +1,58 @@
+GUEST_SH="${2:-wiredtiger-run-ycsb-c5.sh}" # file to run on guest
+QEMU_MEM="${3:-8192}" # memory in MB of QEMU guest
+CORES="${4:-4}" # number of cores to emulate
+EXTRA_DRIVE="${5:-/media/jackson/WDC_HDD/disk_imgs/disk_120G.raw}" # other hard drive to add. TODO: once dev is complete, setup default so that it's not necessary
```
Is it possible to remove this hard-coded path now? I don't like specific user's names in absolute paths.

JacksonDagger replied:
yeah, I'll fix that along with making the extra disk optional

JacksonDagger replied:
it's been removed. by default, there is no extra drive

jnider commented on bench/qemu-test.sh:
```diff
@@ -0,0 +1,58 @@
+
+wait
+
+# there's a typo in wt monitor files
```
Can you give me more details about this typo in wt? Did you look at submitting a patch upstream to fix it?

JacksonDagger replied:
The typo is a missing comma in the csv header which this adds in. I'll take a look at the WT source code and hopefully fix it there

jnider replied:
not 100% necessary, but if it's approximately the same effort for you to fix the WT bug, it will (a) help a lot more people and (b) make your code cleaner - no workaround.

JacksonDagger replied:
Might as well take a look at it. Would be cool to make a (single comma) contribution to WT

JacksonDagger replied:
will leave this in for now and take a look at WT

jnider commented on bench/wiredtiger-copy.sh:
```diff
@@ -0,0 +1,39 @@
+#
+# 2021-08-11 J. Dagger
+#
+# Copy built wiredtiger executables over to QEMU guest
```
Looks like this script assumes the guest is running? If that's true, please add the assumption in the comment. What happens if the guest is not running?

JacksonDagger replied:
I'll add that assumption. Right now, it will build stuff but then hit an error with ssh. I'll add a check to print an error and return an exit code if it doesn't work

JacksonDagger replied:
That assumption in the file comments

jnider commented on bench/wiredtiger-run-ycsb-c4.sh:
```diff
@@ -0,0 +1,18 @@
+cat WT_TEST/monitor
+echo CSV DONE
+
+poweroff -f
```
Why is it important to shut down the machine at the end of the test? This is asymmetric (i.e. you don't start the machine in this script) which can lead to issues when composing scripts later.

JacksonDagger replied:
It's only called from qemu-test.sh which starts the machine. However, I realize now how that implementation could be problematic so I'll try and come up with something better

JacksonDagger replied:
The shutdown is no longer initiated from these scripts. It's now an option in qemu-test.sh

jnider commented on config/rootfs_overlay/root/setup.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
```
What's wrong with bash?

JacksonDagger replied:
It wasn't working on buildroot for me but /bin/sh was. have you been able to get it to work with the bash shebang?

jnider replied:
I haven't tried. But since you mention this is executed in the buildroot environment, it makes sense since buildroot only has busybox which emulates sh and probably doesn't emulate bash (I didn't verify).

jnider commented on swap_setup.sh:
```diff
@@ -1,16 +1,20 @@
 
+IMAGE_SIZE=${1:-1} # set to default of 1GB if not passed in command line
+
+IMAGE_SIZE=${1:-1} # set to default of 1GB if not passed in command line
```
Why are you setting this twice?

JacksonDagger replied:
That's a typo leftover form rebasing. I'll fix it

JacksonDagger committed:
remove duplicate line in swap_setup.sh

JacksonDagger committed:
Update qemu-test.sh structure
qemu-test.sh now has:
-optional host script
-optional guest script
-optional extra drive
-option to choose whether or not to kill after running guest script
run-qemu.sh has been deleted as that function is now effectively
replicated in qemu-test.sh

JacksonDagger committed:
updated file comments

JacksonDagger commented:
@jnider I've made the requested changes and tested them. Please take a look when you have time

JacksonDagger committed:
Benchmarking architecture changes
-Change qemu-test to optargs/flags instead of positional arguments
-New directory structure inside /bench
-Add tool to kill stray QEMU instances inside qemu-test

JacksonDagger committed:
fix typo in filename of ycsb-memconstrict.sh

JacksonDagger commented:
@jnider I've made a few changes. Namely changing to optargs and reworking the directory structure. Lemme know what you think and then we can merge. I'll  put the new memcached stuff on a seperate branch soon

JacksonDagger committed:
Update drive files to not be cached by host. Previously, the swap file would be stored in host memory meaning we weren't getting an accurate look at the performance issues from swapping. See more on QEMU drive caching here https://documentation.suse.com/sles/11-SP4/html/SLES-kvm4zseries/cha-qemu-cachemodes.html

JacksonDagger committed:
Add code for benchmarking outside VM and limiting memory with ramdisks

JacksonDagger committed:
add script for logging memory statistics

JacksonDagger committed:
Merge branch 'benchmarking-scripts' of https://github.com/UBC-ECE-Sasha/PIM-swap into benchmarking-scripts

JacksonDagger committed:
support for local tests

JacksonDagger committed:
added proc monitoring to log mem

JacksonDagger committed:
Add code for parsing local logs

JacksonDagger committed:
update memory logging to convert to kb

JacksonDagger committed:
Merge branch 'benchmarking-scripts' of https://github.com/UBC-ECE-Sasha/PIM-swap into benchmarking-scripts

JacksonDagger committed:
add helper scripts for running n UPMEM machines

JacksonDagger committed:
change some settings and fix typos

JacksonDagger committed:
merge old work

JacksonDagger committed:
add new swap scripts

JacksonDagger committed:
consolidate into one log_mem.sh script in /bench dir for use local and on QEMU

JacksonDagger committed:
streamline wiredtiger on QEMU so there's less config files

JacksonDagger committed:
added basic documentation

JacksonDagger committed:
update local testing script with defaults

JacksonDagger committed:
Merge branch 'master' of https://github.com/UBC-ECE-Sasha/PIM-swap into benchmarking-scripts

JacksonDagger committed:
add scripts for using pidstat and unning wtperf without limiting memory

JacksonDagger committed:
clarofy script hierarchy and make skeleton for log parser

JacksonDagger committed:
seperate limit mem and local-test

JacksonDagger committed:
rename log-mem to standardize dashes vs underscores

JacksonDagger committed:
delete redundant parse_logs.py, add easy script for login to UPMEM machines and update memlimit structure

JacksonDagger committed:
add more descriptive script headers

JacksonDagger committed:
add perf recording

JacksonDagger committed:
fixed bug causing memory fill to fail

JacksonDagger committed:
get rid of log-mem.sh (I was being silly, there are tools that already do this)

JacksonDagger committed:
fix bug in limit-mem that doesn't properly support different ramdisk directories

JacksonDagger committed:
vmstat only display header once

JacksonDagger committed:
parse wiredtiger and vmstat logs

JacksonDagger committed:
have perf record page faults

JacksonDagger committed:
Merge branch 'benchmarking-scripts-dev' of https://github.com/UBC-ECE-Sasha/PIM-swap into benchmarking-scripts-dev

JacksonDagger committed:
seperate logging into own script

JacksonDagger committed:
add descriptions to shell scripts

JacksonDagger committed:
update log parser to read all logs

JacksonDagger committed:
clean up code for removing setup

JacksonDagger committed:
update wtperf configs

JacksonDagger committed:
Merge branch 'benchmarking-scripts-dev' of https://github.com/UBC-ECE-Sasha/PIM-swap into benchmarking-scripts-dev

JacksonDagger committed:
start recording minor and major faults with perf

JacksonDagger committed:
better document scripts

JacksonDagger committed:
add script to plot logs and remove deprecated files

JacksonDagger committed:
add instructions for adding new application

JacksonDagger committed:
Update limit-mem.sh to mount ramdisk by default

JacksonDagger committed:
update process-log.sh to save system state

JacksonDagger committed:
update limit-mem to be less confusing and restore local=test

JacksonDagger committed:
update local-test.sh and add better usage messages

JacksonDagger committed:
fixed some bugs in local-test.sh

JacksonDagger committed:
update to mount with sudo

JacksonDagger committed:
add auto summary to parsed logs and update WT build script

JacksonDagger committed:
Merge branch 'benchmarking-scripts-dev' of https://github.com/UBC-ECE-Sasha/PIM-swap into benchmarking-scripts-dev

JacksonDagger committed:
fix wtperf location

JacksonDagger committed:
add comparison plots and update parse-logs to use timedeltas

JacksonDagger committed:
use inner join and smarten logfile naming when there's no memory limit set

JacksonDagger committed:
umount ramdisk by default

JacksonDagger committed:
add ycsb a and c defaults

JacksonDagger committed:
fix use of sudo

JacksonDagger committed:
add documentation to parse logs

JacksonDagger committed:
Merge branch 'benchmarking-scripts-dev' of https://github.com/UBC-ECE-Sasha/PIM-swap into benchmarking-scripts-dev

JacksonDagger commented:
@shauryapatel1995 @jnider Made a bunch of changes, looking to get this merged in the next week. Let me know if anything is unclear of if there's any changes I can make

jnider commented:
Remind me after the PIM-JPEG paper is done, and I'll take a look

JacksonDagger commented:
> Remind me after the PIM-JPEG paper is done, and I'll take a look

Ok. I'm officially done this position this week but I'll still be around to polish this up and answer questions
